autoload/cmp_nvim_ultisnips.vim

```diff
@@ -1,34 +1,70 @@
 " per snippet) with the keys "trigger", "description", "options" and "value".
 "
 " If 'expandable_only' is 1, only expandable snippets are returned, otherwise all
 " snippets for the current filetype are returned.
 function! cmp_nvim_ultisnips#get_current_snippets(expandable_only)
+  pythonx << EOF
+  import vim
+  from UltiSnips import UltiSnips_Manager, vim_helper
+
+  if vim.eval("a:expandable_only") == "True":
+      before = vim_helper.buf.line_till_cursor
+      snippets = UltiSnips_Manager._snips(before, True)
+  else:
+      snippets = UltiSnips_Manager._snips("", True)
+
+  snippets_info = []
+  vim.command('let g:_cmpu_current_snippets = []')
+  for snippet in snippets:
+      vim.command(
+        "call add(g:_cmpu_current_snippets, {"\
+          "'trigger': pyxeval('str(snippet._trigger)'),"\
+          "'description': pyxeval('str(snippet._description)'),"\
+          "'options': pyxeval('str(snippet._opts)'),"\
+          "'value': pyxeval('str(snippet._value)'),"\
+        "})"
+      )
+  EOF
+  return g:_cmpu_current_snippets
+endfunction
+
+function cmp_nvim_ultisnips#update_fts_get_current_snippets(expandable_only, remove_filetype, additional_filetype)
+  pythonx << EOF
+  import vim
+  from UltiSnips import UltiSnips_Manager, vim_helper
+
+  remove_ft = vim.eval("a:remove_filetype")
+  additional_ft = vim.eval("a:additional_filetype")
+  cur_fts = UltiSnips_Manager._added_buffer_filetypes[vim_helper.buf.number]
+
+  if remove_filetype in cur_fts:
+    cur_fts.remove(remove_ft)
+  if additional_ft not in cur_fts:
+    cur_fts.append(additional_ft)
+  EOF
+  return cmp_nvim_ultisnips#get_current_snippets(expandable_only)
+endfunction
+
+function! cmp_nvim_ultisnips#remove_snippet_filetype(filetype)
 pythonx << EOF
 import vim
 from UltiSnips import UltiSnips_Manager, vim_helper
 
-expandable_only = vim.eval("a:expandable_only")
-if expandable_only == "True":
-    before = vim_helper.buf.line_till_cursor
-    snippets = UltiSnips_Manager._snips(before, True)
-else:
-    snippets = UltiSnips_Manager._snips("", True)
-
-snippets_info = []
-vim.command('let g:_cmpu_current_snippets = []')
-for snippet in snippets:
-    vim.command(
-      "call add(g:_cmpu_current_snippets, {"\
-        "'trigger': pyxeval('str(snippet._trigger)'),"\
-        "'description': pyxeval('str(snippet._description)'),"\
-        "'options': pyxeval('str(snippet._opts)'),"\
-        "'value': pyxeval('str(snippet._value)'),"\
-      "})"
-    )
+ft = vim.eval("a:filetype")
+UltiSnips_Manager._added_buffer_filetypes[vim_helper.buf.number].remove(ft)
 EOF
-return g:_cmpu_current_snippets
+endfunction
+
+function! cmp_nvim_ultisnips#setup_treesitter_autocmds()
+  augroup cmp_nvim_ultisnips
+    autocmd!
+    autocmd TextChangedI * call lua require("cmp_nvim_ultisnips.treesitter").update_ft_at_cursor()
+    autocmd InsertLeave * call lua require("cmp_nvim_ultisnips.treesitter").clear_ft()
+  augroup end
```

Comment:
I'm not sure if we have the performance impact here?

Reply:
Yes I was wondering about this as well. Maybe there is a better way, optimally the function would only be called when the Treesitter node the cursor is on changed.

Reply:
Could you check if you notice any issues with the latest commit?

Reply:
Does it necessary to support it now, for within a file that contains multiple languages, and we should load the snippets for the language at the current cursor?

I see the feature is feasible to do, and great work that you did. Congrats

But it's also a tradeoff, we did yet notice as of now. But it's not performance efficient since it continuously calls the event when the cursor moves between every single character in Insert mode. Because in a single line, there would be just a single language for it. We could improve to call the event when the cursor enters on the new line but it's way more complicated.

That was my thought. You can think about it and could take no action for it. Anyway, your work is greate. Thanks for your help in making this plugin fantastic.